File: libraries/analytics/src/main/java/com/example/analytics/AnalyticsUtils.java
package com.example.analytics;

import android.app.Activity;
import android.content.Context;
import android.os.Bundle;
import android.support.annotation.NonNull;
import android.support.annotation.Nullable;

import com.google.firebase.analytics.FirebaseAnalytics;

public class AnalyticsUtils {
	public static void trackAction(@NonNull Context context, @NonNull @Action String action) {
		trackAction(context, action, null);
	}

	public static void trackAction(@NonNull Context context, @NonNull @Action String action, @Nullable Bundle extras) {
		FirebaseAnalytics.getInstance(context).logEvent(action, extras);
	}

	public static void trackView(@NonNull Activity activity, @NonNull @Screen String screen) {
		FirebaseAnalytics.getInstance(activity).setCurrentScreen(activity, screen, null);
	}
}

File: analytics/src/main/kotlin/br/com/vitorsalgado/example/analytics/AnalyticsUtils.kt
package br.com.vitorsalgado.example.analytics

import android.app.Activity
import android.content.Context
import java.util.*

object AnalyticsUtils {
  private val subscriber = mutableListOf<AnalyticsAdapter>()

  fun attach(observer: AnalyticsAdapter) {
    if (subscriber.contains(observer))
      throw IllegalArgumentException("observer of type ${observer.javaClass.simpleName} already added")

    subscriber.add(observer)
  }

  fun detach(observer: AnalyticsAdapter) = subscriber.remove(observer)

  fun detachAll() = subscriber.clear()

  fun trackAction(context: Context, action: String) =
    trackAction(
      context,
      TraceActionArgs(action, Collections.emptyMap())
    )

  fun trackAction(context: Context, args: TraceActionArgs) =
    subscriber.forEach { it.trackAction(context, args) }

  fun trackView(context: Context, args: TraceScreenArgs) =
    subscriber.forEach { it.trackView(context, args) }
}
